pool.js

```diff
@@ -296,32 +323,91 @@ Pool.prototype.request_count = function () {
     return count;
 };
 
-Pool.prototype.close = function () {
+Pool.prototype.close = function close() {
     var endpoints = this.endpoints;
     for (var i = 0; i < endpoints.length; i++) {
         endpoints[i].close();
     }
 };
 
-// Dynamic membership
-Pool.prototype.add_endpoint = function (host_port) {
+Pool.prototype.valid_host_port = function valid_host_port(host_port) {
     var ip_port = host_port.split(":");
     var ip = ip_port[0];
     var port = +ip_port[1];
     if (port > 0 && port < 65536) {
-        var endpoint = new GO.PoolEndpoint(this.http, ip, port, this.options);
-        endpoint.on("health", this.endpoint_health_changed.bind(this));
-        endpoint.on("timeout", this.endpoint_timed_out.bind(this));
-        this.endpoints.push(endpoint);
-        this.length++;
-        this.endpoints_by_name[host_port] = endpoint;
+        return ip_port;
     }
+    return null;
 };
 
-Pool.prototype.remove_endpoint = function (host_port) {
-    var ip_port = host_port.split(":");
+// Dynamic membership
+Pool.prototype.add_endpoint = function add_endpoint(host_port) {
+    var ip_port = this.valid_host_port(host_port);
+    if (!ip_port) {
+        return;
+    }
+
+    var hasHostPort = false;
+    for (var i = 0; i < this.all_hostports.length; i++) {
+        if (this.all_hostports[i] === host_port) {
+            hasHostPort = true;
+            break;
+        }
+    }
+
+    if (!hasHostPort) {
+        this.all_hostports.push(host_port);
+    }
+
+    var maxSizeIsRelevant = this.max_pool_size &&
+        this.max_pool_size < this.all_hostports.length;
+
+    // if using maxSize, probabilistically add the new host port to the
+    // active pool endpoints
+    if (maxSizeIsRelevant) {
+        var r = Math.random() * this.all_hostports.length;
+        if (r < this.max_pool_size) {
+            this.add_pool_endpoint(host_port);
```

Comment:
I don't entirely follow the "fully connected pool" example.  "unused" list only factors in to adjust_pool_size when the adjustment is to add,  not when removing, as is the case here. maybe we can talk that out IRL.

This add then remove is a little convoluted because I was trying to optimize allocation ... decide early that it's not included and not init  PoolEndpoint.
anyway you're right about this being a bit biased.  here is a deeper analysis and potential solution.   

before adding
m = max_pool_size
n =  all_host 

begin add_endpoint
n increases to  n+1

P(added)  = P(step1: added) * P(step2: not removed)
P(added)  =  m/(n+1) *  (1 - 1/(m+1)) =  m/(n+1) * m/(m+1)

This illustrates the problem , as we see for small values of m.  the P(added) will biased low.
for  m=1, n=2 for example
P(added) = 1/3 * 1/2 =  1/6      <--  bad  there are 3 IPs

Ideally  we want the new ip to be added with probability of just m/(n+1).    So to correct  we can scale the  P(step1) by (m+1)/m.
   P(step1: added)  =  (m/n+1) * (m+1)/m  =   (m+1)/(n+1)

with bias fix
P(added) =  2/3 * 1/2 = 1/3     <--  correct

So i think some careful additions of +1 s to end of  lines 368, 369   should resolve the bias.